#4519: Adding ability to configure custom Authorizer implementation (#4540)
strimzi/strimzi-kafka-operator: documentation/api/io.strimzi.api.kafka.model.KafkaAuthorizationCustom.adoc

ACLs allow you to define which users have access to which resources at a granular level.

Configure the `Kafka` custom resource to use custom authorization.
Set the `type` property in the `authorization` section to the value `custom`,
and the set following properties.

IMPORTANT: The custom Authorizer must implement the org.apache.kafka.server.authorizer.Authorizer interface, and support configuration of super users using the super.users configuration property.

=== `authorizerClass`
(Required)Java class that implements the `org.apache.kafka.server.authorizer.Authorizer` interface to support the custom ACLs.

[id='property-simple-authorization-superusers-{context}']
=== `superUsers`
A list of user principals treated as super users, so that they are always allowed without querying ACL rules.
For more information see xref:con-securing-kafka-authorization-str[Kafka authorization].

Configuration required for the custom authorizer can be provided using `Kafka.spec.kafka.config` and those properties will be supplied to custom authorizer in `config` method for initialization.

.An example of custom authorization configuration under `Kafka.spec`
[source,yaml,subs="attributes+"]
----

> authorization:
>   type: custom
>   authorizerClass: io.mycompany.CustomAuthorizer
>   superUsers:
>     - CN=client_1
>     - user_2
>     - CN=client_3
> config:
>   strimzi.authorization.custom.property1=value1
>   strimzi.authorization.custom.property2=value2

Reviewer: I wonder if this makes it a bit easier to read.

```suggestion
spec:
  kafka:
    # ...
    authorization:
      type: custom
      authorizerClass: io.mycompany.CustomAuthorizer
      superUsers:
        - CN=client_1
        - user_2
        - CN=client_3
    # ...
    config:
      authorization.custom.property1=value1
      authorization.custom.property2=value2
    # ...
```

    public List<? extends CompletionStage<AclDeleteResult>> deleteAcls(AuthorizableRequestContext requestContext,
            List<AclBindingFilter> aclBindingFilters) {
        // TODO: your custom implementation
    }

    @Override
    public Iterable<AclBinding> acls(AclBindingFilter filter) {
        // TODO: your custom implementation
    }

    public static class UserSpec {
        private final String type;
        private final String name;

        private UserSpec(String type, String name) {
            this.type = type;
            this.name = name;
        }

        public String getType() {
            return type;
        }

        public String getName() {
            return name;
        }

        public static UserSpec of(String principal) {
            int pos = principal.indexOf(':');
            if (pos <= 0) {
                throw new IllegalArgumentException("Invalid user specification: " + principal);
            }
            return new UserSpec(principal.substring(0, pos), principal.substring(pos + 1));
        }

        @Override
        public String toString() {
            return super.toString() + " " + type + ":" + name;
        }
    }
}

> ----

Reviewer: I not sure this example makes sense ... the docs would need to be checked and maintained when the Authorizer interface ever changes. Maybe we should leave it to Kafka.

Author: Yes, agree. Will remove it

> Custom authorization in Strimzi lets user configure their own `Authorizer` plugin to define the Access Control Lists (ACLs)

Reviewer 2: ```suggestion
To use custom authorization in Strimzi, you can configure your own `Authorizer` plugin to define Access Control Lists (ACLs).
```

ACLs allow you to define which users have access to which resources at a granular level.

Configure the `Kafka` custom resource to use custom authorization.
Set the `type` property in the `authorization` section to the value `custom`,

> and the set following properties.

Reviewer: Can we add here something like ...

```
IMPORTANT: The custom authorizer needs to implement the `org.apache.kafka.server.authorizer.Authorizer` interface and support configuration of super-users from the `super.users` configuration field.
```

or something like that? This is important, because this is needed to configure the Strimzi compoenents.

Reviewer 2: IMPORTANT: The custom `Authorizer` must implement the `org.apache.kafka.server.authorizer.Authorizer` interface, and support configuration of super users using the `super.users` configuration property.

=== `implementionClass`
Implementation class of the `org.apache.kafka.server.authorizer.Authorizer` interface to support the custom ACLs
**Required.**

[id='property-simple-authorization-superusers-{context}']
=== `superUsers`
A list of user principals treated as super users, so that they are always allowed without querying ACL rules.
For more information see xref:con-securing-kafka-authorization-str[Kafka authorization].

> .An example of simple authorization configuration

Reviewer 2: ```suggestion
.An example of custom authorization configuration
```

> To use custom authorization in Strimzi, you can configure your own `Authorizer` plugin to define Access Control Lists (ACLs).

Reviewer: Can we also (somewhere here) add some explanation that users need to build their custom image and add the custom authorizer JARs to the classpath (`/opt/kafka/libs`)? Otherwise some might expect that it would be magically added just by using the API in the Kafka CR ;-).

Author: how about adding the below text? @scholzj @PaulRMellor 

For Strimzi to instantiate and configure the custom Authorizer class, apart from the additions to the custom resource as shown above, the JAR file containing the custom Authorizer class along with its dependencies must be available on the classpath of Kafka broker.

Strimzi maven build process provides a mechanism to add these custom third-party libraries to the generated Kafka broker container image by adding them as dependencies in the `pom.xml` file under `docker-images/kafka/kafka-thirdparty-libs` directory. Under this directory there are different folders for different Kafka versions, choose the folder that fits your needs. Before `pom.xml` is modified the third-party library must be available in a Maven repository, and that maven repository is accessible to the Strimzi build process.

Reviewer: That sounds good to me! 👍

To use custom authorization in Strimzi, you can configure your own `Authorizer` plugin to define Access Control Lists (ACLs).

ACLs allow you to define which users have access to which resources at a granular level.

Configure the `Kafka` custom resource to use custom authorization.
Set the `type` property in the `authorization` section to the value `custom`,
and the set following properties.

> IMPORTANT: The custom Authorizer must implement the org.apache.kafka.server.authorizer.Authorizer interface, and support configuration of super users using the super.users configuration property.

Reviewer 2: ```suggestion
IMPORTANT: The custom Authorizer must implement the `org.apache.kafka.server.authorizer.Authorizer` interface, and support configuration of super users using the super.users configuration property.
```

[id='property-custom-authorization-authorizerclass-{context}']
=== `authorizerClass`

> (Required) Java class that implements the `org.apache.kafka.server.authorizer.Authorizer` interface to support the custom ACLs.

Reviewer 2: ```suggestion
(Required) Java class that implements the `org.apache.kafka.server.authorizer.Authorizer` interface to support custom ACLs.
```

[id='property-custom-authorization-superusers-{context}']
=== `superUsers`
A list of user principals treated as super users, so that they are always allowed without querying ACL rules.
For more information see xref:con-securing-kafka-authorization-str[Kafka authorization].

> Configuration required for the custom authorizer can be provided using `Kafka.spec.kafka.config` and those properties will be supplied to custom authorizer in `config` method for initialization.

Reviewer 2: ```suggestion
You can add configuration for initializing the custom authorizer using `Kafka.spec.kafka.config`.
```

.An example of custom authorization configuration under `Kafka.spec`
[source,yaml,subs="attributes+"]
----
spec:
  kafka:
    # ...
    authorization:
      type: custom
      authorizerClass: io.mycompany.CustomAuthorizer
      superUsers:
        - CN=client_1
        - user_2
        - CN=client_3
    # ...
    config:
      authorization.custom.property1=value1
      authorization.custom.property2=value2
    # ...
----

> For Strimzi to instantiate and configure the custom Authorizer class, apart from the additions to the custom resource as shown above, the JAR file containing the custom Authorizer class along with its dependencies must be available on the classpath of Kafka broker.

Reviewer 2: ```suggestion
In addition to the `Kafka` custom resource configuration, the JAR file containing the custom authorizer class along with its dependencies must be available on the classpath of the Kafka broker. 
```

> Strimzi maven build process provides a mechanism to add these custom third-party libraries to the generated Kafka broker container image by adding them as dependencies in the pom.xml file under docker-images/kafka/kafka-thirdparty-libs directory. Under this directory there are different folders for different Kafka versions, choose the folder that fits your needs. Before pom.xml is modified the third-party library must be available in a Maven repository, and that maven repository is accessible to the Strimzi build process.

Reviewer 2: ```suggestion
The Strimzi Maven build process provides a mechanism to add custom third-party libraries to the generated Kafka broker container image by adding them as dependencies in the `pom.xml` file under the `docker-images/kafka/kafka-thirdparty-libs` directory. The directory contains different folders for different Kafka versions. Choose the appropriate folder. Before modifying the `pom.xml` file, the third-party library must be available in a Maven repository, and that Maven repository must be accessible to the Strimzi build process.
```
